### PR TITLE
nixos/db-rest: init

### PR DESCRIPTION
**Pull request by NyCodeGHG**
## Description of changes

added nixos module for db-rest

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/misc/db-rest.nix
Comment (h7x4) [inline nixos/modules/services/misc/db-rest.nix:88]: Maybe something like this would be more accurate?

```suggestion
        assertion = (cfg.redis.enable && !cfg.redis.createLocally) -> (cfg.redis.host != null && cfg.redis.port != null);
```

File: nixos/modules/services/misc/db-rest.nix
Comment (h7x4) [inline nixos/modules/services/misc/db-rest.nix:90]: These docs seems to be formatted as docbook. This was deprecated in favor of commonmark about a year ago or something.

You can read more about it here if you'd like: https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#syntax

```suggestion
          {option}`services.db-rest.redis.createLocally` and redis network connection ({option}`services.db-rest.redis.host` or {option}`services.db-rest.redis.port`) enabled. Disable either of them.
```

File: nixos/modules/services/misc/db-rest.nix
Comment (h7x4) [inline nixos/modules/services/misc/db-rest.nix:184]:
```suggestion
    services.redis.servers.db-rest.enable = cfg.redis.enable && cfg.redis.createLocally;
```

**Review by h7x4 — COMMENTED** [review]
Here's a few comments

File: nixos/modules/services/misc/db-rest.nix
Comment (NyCodeGHG) [reply nixos/modules/services/misc/db-rest.nix:90]: Oh TIL thanks, i copied this from *somewhere* 
i don't remember

File: nixos/modules/services/misc/db-rest.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/db-rest.nix:100]: I don't think we are doing this anywhere else and sure it is insecure but we should probably fix this on a wider scope by adding a path type which forbids this instead of adding asserts to single modules.

File: nixos/modules/services/misc/db-rest.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/db-rest.nix:159]: This should use the systemd.package option.

File: nixos/modules/services/misc/db-rest.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/db-rest.nix:165]: This can be entirely done in Environment as far as I can tell.

File: nixos/modules/services/misc/db-rest.nix
Comment (SuperSandro2000) [inline nixos/modules/services/misc/db-rest.nix:166]: and then we can move this to ExecStart

File: nixos/modules/services/misc/db-rest.nix
Comment (SuperSandro2000) [reply nixos/modules/services/misc/db-rest.nix:165]: Oh, I think that does not work.

**Comment by NyCodeGHG** [comment]
Anything else missing to get this merged?

**Comment by jeyemwey** [comment]
Hi, is there anything that is still blocking this PR?

**Comment by lilyinstarlight** [comment]
@ofborg test db-rest

File: nixos/modules/services/misc/db-rest.nix
Comment (lilyinstarlight) [inline nixos/modules/services/misc/db-rest.nix:40]:
```suggestion
        enable = mkEnableOption "caching with redis for db-rest.";
```

File: nixos/modules/services/misc/db-rest.nix
Comment (lilyinstarlight) [inline nixos/modules/services/misc/db-rest.nix:161]:
```suggestion
              export REDIS_URL="${protocol}://${username}:$(${lib.getExe' config.systemd.package "systemd-creds"} cat REDIS_PASSWORD)@${host}:${toString port}"
```

I do wonder if this can be done any better (especially with respect to escaping), but this is adequate since it won't expose secret on any command lines (do at least make the `toString` change though since this would currently fail)

File: nixos/modules/services/misc/db-rest.nix
Comment (lilyinstarlight) [inline nixos/modules/services/misc/db-rest.nix:162]:
```suggestion
              exec ${lib.escapeShellArgs (lib.getExe cfg.package)}
```

Suggestion is slightly safer/more robust, but this is fine to be honest

File: nixos/modules/services/misc/db-rest.nix
Comment (lilyinstarlight) [inline nixos/modules/services/misc/db-rest.nix:117]:
```suggestion
          ExecStart = mkDefault lib.getExe cfg.package;
```

Slightly more robust, but not required

File: nixos/tests/db-rest.nix
Comment (lilyinstarlight) [inline nixos/tests/db-rest.nix:99]: Maybe ensure `curl` is in `environment.systemPackages`? Not sure how much it realisticially matters

**Review by lilyinstarlight — COMMENTED** [review]
~~The `toString` part of the second review item is probably required.~~ Otherwise this is adequate for me to merge

Edit: I misread, this is good to go!

**Review by lilyinstarlight — APPROVED** [review]

**Pull request merged**
